Add dithering to the `Speech2TextFeatureExtractor` API.

KarelVesely84 opened the pull request:
Enable dithering in speech-to-text feature extraction. Dithering exists in the original kaldi features.
- in kaldi : https://github.com/kaldi-asr/kaldi/blob/4a8b7f673275597fef8a15b160124bd0985b59bd/src/feat/feature-window.cc#L145

The dithering is adding small Gaussian noise to the waveform on input of feature extraction.
This is helpful for audio signals with hard-zero sections due to HW VAD, these hard-zeros
may break the ASR training or inference if they appear in the data.

With dithering without a seed, the features become non-deterministic due to small Gaussian noise added to the audio (i.e. 2 runs lead to little different outputs). When debugging feature extraction code, it is good to set dithering to 0.0 (i.e. default value).

KarelVesely84 committed:
Add dithering to the `Speech2TextFeatureExtractor` API.

- in kaldi : https://github.com/kaldi-asr/kaldi/blob/4a8b7f673275597fef8a15b160124bd0985b59bd/src/feat/feature-window.cc#L145
- with dithering without a seed, the features become non-deterministic due
  to small Gaussian noise added to the audio (i.e. 2 runs lead to little
  different outputs)

KarelVesely84 commented:
Hello, is there somebody to look into this for a review ?
Thank you,
K.V.

LysandreJik commented:
cc @ylacombe @eustlb @Vaibhavs10 regarding this proposed feature contribution

Vaibhavs10 reviewed (commented):
I'd be supportive of this generally. For more context @KarelVesely84 - How are you using this in training with Kaldi and for which models? Are you using the example training/ fine-tuning script?

ylacombe commented on src/transformers/audio_utils.py:
```diff
@@ -460,6 +461,9 @@ def spectrogram(
             If True, only computes the positive frequencies and returns a spectrogram containing `fft_length // 2 + 1`
             frequency bins. If False, also computes the negative frequencies and returns `fft_length` frequency bins.
+        dither (`float`):
+            Add dithering (add small Gaussian noise to each frame).
+            E.g. use 4 to add dithering, 0.0 means no dithering.
```
(nit)
```suggestion
            Adds dithering. In other words, adds a small Gaussian noise to each frame.
            E.g. use 4.0 to add dithering with a normal distribution centered around 0.0 with standard deviation 4.0. 0.0 means no dithering.
```

ylacombe commented on src/transformers/audio_utils.py:
```diff
@@ -653,6 +661,9 @@ def spectrogram_batch(
         onesided (`bool`, *optional*, defaults to `True`):
             If True, returns a one-sided spectrogram for real input signals.
+        dither (`float`):
+            Add dithering (add small Gaussian noise to each frame).
```
Same suggestion as above here and for the rest of the PR!

ylacombe reviewed (approved):
Hey @KarelVesely84, thanks for adding this, it looks great to me!

You could maybe add a small test to the feature extraction tests of speech2text, just in case, to ensure that we get the expected results with a set seed.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_34638). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

KarelVesely84 committed:
update the PR

- add dithering also for WhisperFeatureExtractor
- not adding to Wav2Vec2FeatureExtractor (no FBANK computation)

KarelVesely84 commented:
Hello,
here is the correspoding example code, it does inference from existing models:
https://github.com/KarelVesely84/transformers_sandbox/tree/main/dithering

It is a dummy example, so training is without dithering and inference is with dithering.
The worsening results demonstrate that dithering is happening.
K.

KarelVesely84 commented:
I see the [feature extraction tests](https://github.com/huggingface/transformers/blob/main/tests/models/speech_to_text/test_feature_extraction_speech_to_text.py), but how sholud the test actually look like ? 
T1: Compute features with / without dithering and make sure the outputs are different, and not "too different" ?

KarelVesely84 commented:
Hello @ylacombe , the feedback was integrated into the PR. Is there something more to do ?
Best regards,
Karel

ylacombe reviewed (approved):
LGTM!

> I see the [feature extraction tests](https://github.com/huggingface/transformers/blob/main/tests/models/speech_to_text/test_feature_extraction_speech_to_text.py), but how sholud the test actually look like ?
T1: Compute features with / without dithering and make sure the outputs are different, and not "too different" ?

Exactly! Let's do this before merging

ylacombe commented on src/transformers/audio_utils.py:
```diff
@@ -460,6 +461,10 @@ def spectrogram(
         onesided (`bool`, *optional*, defaults to `True`):
             If True, only computes the positive frequencies and returns a spectrogram containing `fft_length // 2 + 1`
             frequency bins. If False, also computes the negative frequencies and returns `fft_length` frequency bins.
+        dither (`float`):
```
You should also make sure that the docstrings is right here and in the rest of the docstrings
```suggestion
        dither (`float`, *optional*, defaults to 0.0):
```

KarelVesely84 committed:
add unit-tests for dithering, fix docstrings

KarelVesely84 committed:
ruff

KarelVesely84 committed:
utils/check_copies.py --fix_and_overwrite

KarelVesely84 committed:
Merge remote-tracking branch 'origin/main' into add_dither

KarelVesely84 commented:
Ok, it seems to be ready.

ylacombe commented on src/transformers/audio_utils.py:
```diff
@@ -540,6 +545,9 @@ def spectrogram(
         buffer[:frame_length] = waveform[timestep : timestep + frame_length]
 
+        if dither != 0.0:
+            buffer[:frame_length] += dither * np.random.randn(*buffer[:frame_length].shape)
```
According to the docstrings, waveform is supposed to be a 1D signal. In that case, it should be simpler to do it as proposed?
 ```suggestion
             buffer[:frame_length] += dither * np.random.randn(frame_length)
```

Or maybe `len(buffer)` ?

ylacombe commented on tests/models/clap/test_feature_extraction_clap.py:
```diff
@@ -53,7 +53,7 @@ def floats_list(shape, scale=1.0, rng=None, name=None):
 @require_torch
 @require_torchaudio
 # Copied from tests.models.whisper.test_feature_extraction_whisper.WhisperFeatureExtractionTester with Whisper->Clap
-class ClapFeatureExtractionTester(unittest.TestCase):
```
Why did you remove this part, out of curiosity?

ylacombe commented on tests/models/speech_to_text/test_feature_extraction_speech_to_text.py:
```diff
@@ -144,6 +144,38 @@ def test_call(self):
             self.assertTrue(np.allclose(enc_seq_1, enc_seq_2, atol=1e-3))
 
+    def test_dither(self):
+        # Tests that features with and without little dithering are similar, but not the same
```
Let's set the seed here, to ensure reproducibility.

ylacombe commented on tests/models/speech_to_text/test_feature_extraction_speech_to_text.py:
```diff
@@ -144,6 +144,38 @@ def test_call(self):
+        # features are not identical
+        self.assertTrue(np.abs(diff).mean() > 1e-5)
+        # features are not too different
+        self.assertTrue(np.abs(diff).mean() <= 1e-3)
+        self.assertTrue(np.abs(diff).max() <= 1e-2)
```
Great!

ylacombe commented on src/transformers/audio_utils.py:
```diff
@@ -653,6 +662,10 @@ def spectrogram_batch(
             The padding strategy when `center` is `True`.
         onesided (`bool`, *optional*, defaults to `True`):
             If True, returns a one-sided spectrogram for real input signals.
+        dither (`float`):
```
```suggestion
        dither (`float`, *optional*, defaults to 0.0):
```

ylacombe reviewed (approved):
LGTM! Thanks for iterating

@ArthurZucker, could you review when you have time? Thanks!

KarelVesely84 committed:
update code, add seed to unit-test

KarelVesely84 committed:
Merge branch 'main' into add_dither

KarelVesely84 commented:
Hello, is there something other to be done ?
Best regards,
K. Vesely

Vaibhavs10 reviewed (commented):
Yoach already reviewed it from Audio side, I think we're only looking for a core-maintainer review + merge now: https://github.com/huggingface/transformers/pull/34638#pullrequestreview-2488328459

Rocketknight1 commented:
Got it - pinging @arthurzucker @LysandreJik for core maintainer review!

ArthurZucker commented on src/transformers/audio_utils.py:
```diff
@@ -460,6 +461,10 @@ def spectrogram(
             If True, only computes the positive frequencies and returns a spectrogram containing `fft_length // 2 + 1`
             frequency bins. If False, also computes the negative frequencies and returns `fft_length` frequency bins.
+        dither (`float`, *optional*, defaults to 0.0):
+            Adds dithering. In other words, adds a small Gaussian noise to each frame.
+            E.g. use 4.0 to add dithering with a normal distribution centered
+            around 0.0 with standard deviation 4.0, 0.0 means no dithering.
```
let's add the comment about "this can help for hard audio in ASR" 😉

ArthurZucker reviewed (approved):
Thanks, would be nice to just say why people should use this in the doc!

KarelVesely84 committed:
adding explanation of dithering

KarelVesely84 committed:
Merge remote-tracking branch 'origin/main' into add_dither

ArthurZucker commented:
could you resolve the last conflicts and we should be good to merge!

KarelVesely84 committed:
Merge remote-tracking branch 'origin/main' into add_dither

KarelVesely84 commented:
> could you resolve the last conflicts and we should be good to merge!

ok, done

ArthurZucker commented:
Thanks for the PR! 🤗

mizoru commented:
@KarelVesely84 
> these hard-zeros may break the ASR training or inference if they appear in the data
I encountered this problem with Whisper hallucinating on hard zeros. Could you point me in a direction where it is explained how these lead to OOD data? I thought it was caused by normalization at first, but I can't tell the difference between dithered and non-dithered spectrograms when I visualize them. So how do they manage to trip up models?
